File: src/stores/items.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import { invoke } from '@tauri-apps/api/core'
import { useTagsStore } from '@/stores/tags'
import type { Item, Tag } from '@/types'

export const useItemsStore = defineStore('items', () => {
  const items = ref<Item[]>([])
  const loading = ref(false)
  const error = ref<string | null>(null)

  async function createItem(
    path: string,
    isDirectory: boolean,
    size: number | null = null,
    modifiedTime: number | null = null
  ) {
    try {
      const id = await invoke<number>('create_item', {
        path,
        isDirectory,
        size,
        modifiedTime,
      })
      return id
    } catch (e) {
      error.value = e as string
      console.error('Failed to create item:', e)
      throw e
    }
  }

  async function getItem(id: number) {
    try {
      const item = await invoke<Item>('get_item', { id })
      return item
    } catch (e) {
      error.value = e as string
      console.error('Failed to get item:', e)
      throw e
    }
  }

  async function getItemByPath(path: string) {
    try {
      const item = await invoke<Item | null>('get_item_by_path', { path })
      return item
    } catch (e) {
      error.value = e as string
      console.error('Failed to get item by path:', e)
      throw e
    }
  }

  async function getItemsByPaths(paths: string[]) {
    try {
      const items = await invoke<Item[]>('get_items_by_paths', { paths })
      return items
    } catch (e) {
      error.value = e as string
      console.error('Failed to get items by paths:', e)
      throw e
    }
  }

  async function addTagToItem(itemId: number, tagId: number) {
    try {
      await invoke('add_tag_to_item', {
        itemId,
        tagId,
      })
      const tagsStore = useTagsStore()
      await tagsStore.loadUsageCounts()
      tagsStore.itemTagsVersion++
    } catch (e) {
      error.value = e as string
      console.error('Failed to add tag to item:', e)
      throw e
    }
  }

  async function removeTagFromItem(itemId: number, tagId: number) {
    try {
      await invoke('remove_tag_from_item', {
        itemId,
        tagId,
      })
      const tagsStore = useTagsStore()
      await tagsStore.loadUsageCounts()
      tagsStore.itemTagsVersion++
    } catch (e) {
      error.value = e as string
      console.error('Failed to remove tag from item:', e)
      throw e
    }
  }

  async function getTagsForItem(itemId: number) {
    try {
      const tags = await invoke<Tag[]>('get_tags_for_item', { itemId })
      return tags
    } catch (e) {
      error.value = e as string
      console.error('Failed to get tags for item:', e)
      throw e
    }
  }

  async function getTagsForItems(itemIds: number[]) {
    try {
      const tagsMap = await invoke<Record<number, Tag[]>>('get_tags_for_items', { itemIds })
      return tagsMap
    } catch (e) {
      error.value = e as string
      console.error('Failed to get tags for items:', e)
      throw e
    }
  }

  async function updateItemTags(itemId: number, tagIds: number[]) {
    try {
      await invoke('update_item_tags', { itemId, tagIds })
      const tagsStore = useTagsStore()
      await tagsStore.loadUsageCounts()
      tagsStore.itemTagsVersion++
    } catch (e) {
      error.value = e as string
      console.error('Failed to update item tags:', e)
      throw e
    }
  }

  return {
    items,
    loading,
    error,
    createItem,
    getItem,
    getItemByPath,
    getItemsByPaths,
    addTagToItem,
    removeTagFromItem,
    getTagsForItem,
    getTagsForItems,
    updateItemTags,
  }
})

File: src-tauri/src/commands/items.rs
//! Item Commands
//!
//! Thin adapters for item operations that delegate to ItemService.

use crate::application::dto::{CreateItemDto, ItemDto, TagDto, UpdateItemDto};
use crate::error::{AppError, AppResult};
use crate::state::AppState;
use std::collections::HashMap;
use tauri::State;

#[tauri::command]
pub async fn create_item(
    path: String,
    is_directory: bool,
    size: Option<i64>,
    modified_time: Option<i64>,
    state: State<'_, AppState>,
) -> AppResult<i64> {
    let dto = CreateItemDto {
        path,
        is_directory,
        size,
        modified_time,
    };

    state
        .item_service
        .create(dto)
        .await
        .map_err(|e| AppError::InvalidInput(e.to_string()))
}

#[tauri::command]
pub async fn get_item(id: i64, state: State<'_, AppState>) -> AppResult<ItemDto> {
    state
        .item_service
        .get_by_id(id)
        .await
        .map_err(|e| AppError::InvalidInput(e.to_string()))?
        .ok_or_else(|| AppError::NotFound(format!("Item with id {}", id)))
}

#[tauri::command]
pub async fn get_item_by_path(
    path: String,
    state: State<'_, AppState>,
) -> AppResult<Option<ItemDto>> {
    state
        .item_service
        .get_by_path(&path)
        .await
        .map_err(|e| AppError::InvalidInput(e.to_string()))
}

#[tauri::command]
pub async fn get_items_by_paths(
    paths: Vec<String>,
    state: State<'_, AppState>,
) -> AppResult<Vec<ItemDto>> {
    state
        .item_service
        .get_by_paths(paths)
        .await
        .map_err(|e| AppError::InvalidInput(e.to_string()))
}

#[tauri::command]
pub async fn update_item(
    id: i64,
    path: Option<String>,
    size: Option<i64>,
    modified_time: Option<i64>,
    state: State<'_, AppState>,
) -> AppResult<()> {
    let dto = UpdateItemDto {
        path,
        size,
        modified_time,
    };

    state
        .item_service
        .update(id, dto)
        .await
        .map_err(|e| AppError::InvalidInput(e.to_string()))
}

#[tauri::command]
pub async fn delete_item(id: i64, state: State<'_, AppState>) -> AppResult<()> {
    state
        .item_service
        .delete(id)
        .await
        .map_err(|e| AppError::NotFound(e.to_string()))
}

#[tauri::command]
pub async fn add_tag_to_item(
    item_id: i64,
    tag_id: i64,
    state: State<'_, AppState>,
) -> AppResult<()> {
    state
        .item_service
        .add_tag(item_id, tag_id)
        .await
        .map_err(|e| AppError::InvalidInput(e.to_string()))
}

#[tauri::command]
pub async fn remove_tag_from_item(
    item_id: i64,
    tag_id: i64,
    state: State<'_, AppState>,
) -> AppResult<()> {
    state
        .item_service
        .remove_tag(item_id, tag_id)
        .await
        .map_err(|e| AppError::InvalidInput(e.to_string()))
}

#[tauri::command]
pub async fn get_tags_for_item(item_id: i64, state: State<'_, AppState>) -> AppResult<Vec<TagDto>> {
    state
        .item_service
        .get_tags(item_id)
        .await
        .map_err(|e| AppError::InvalidInput(e.to_string()))
}

#[tauri::command]
pub async fn get_tags_for_items(
    item_ids: Vec<i64>,
    state: State<'_, AppState>,
) -> AppResult<HashMap<i64, Vec<TagDto>>> {
    state
        .item_service
        .get_tags_batch(item_ids)
        .await
        .map_err(|e| AppError::InvalidInput(e.to_string()))
}

#[tauri::command]
pub async fn update_item_tags(
    item_id: i64,
    tag_ids: Vec<i64>,
    state: State<'_, AppState>,
) -> AppResult<()> {
    state
        .item_service
        .update_tags(item_id, tag_ids)
        .await
        .map_err(|e| AppError::InvalidInput(e.to_string()))
}
